distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Range.enso

```diff
@@ -0,0 +1,484 @@
+         into a new date.
+       - if_empty: Value returned if the range is empty.
+    reduce : (Date -> Date -> Date) -> Any -> Any
+    reduce self function ~if_empty=(Error.throw Empty_Error) =
+        case self.length of
+            0 -> if_empty
+            1 -> self.start
+            length ->
+                (1.up_to length).fold self.start acc-> ix->
+                    function acc (self.internal_at ix)
+
+
+## PRIVATE
+   Computes the length of the range and verifies its invariants.
+
+   If any of the invariants are violated, a dataflow error is raised.
+compute_length_and_verify : Date -> Date -> Period -> Boolean -> Integer ! Illegal_Argument
+compute_length_and_verify start end step increasing =
+    if is_period_positive step . not then Error.throw (Illegal_Argument.Error "The step `Period` for `Date_Range` must be positive, i.e. all of `years`, `months` and `days` must be non-negative and at least one of them must be strictly positive.") else
+        is_range_empty = case increasing of
+            True -> start >= end
+            False -> start <= end
+        if is_range_empty then 0 else
+            # First a few heuristics for a fast path.
+            # If there are no years or months, we can perform a simple computation on day difference.
+            if step.total_months == 0 then compute_length_step_days start end step.days increasing else
+                # Similarly, if we are only shifting by months, we can rely on a simpler computation.
+                if step.days == 0 then compute_length_step_months start end step.total_months increasing else
+                    # Then we go brute force for the general case.
+                    compute_length_step_brute_force start end step increasing
+
+## PRIVATE
+is_period_positive period =
+    if (period.years < 0) || (period.months < 0) || (period.days < 0) then False else
+        (period.total_months > 0) || (period.days > 0)
+
+## PRIVATE
+   Assumes that the range is not empty.
+compute_length_step_days : Date -> Date -> Integer -> Boolean -> Integer
+compute_length_step_days start end step increasing =
+    # Logic analogous to `Range.length`.
+    diff = case increasing of
+        True  -> Time_Utils.days_between start end
+        False -> Time_Utils.days_between end start
+    # assert (diff >= 0)
+    steps = diff . div step
+    exact_fit = diff % step == 0
+    if exact_fit then steps else steps+1
+
+## PRIVATE
+   Assumes that the range is not empty.
+compute_length_step_months start end step increasing =
+    diff = case increasing of
+        True  -> Time_Utils.months_between start end
+        False -> Time_Utils.months_between end start
+    # assert (diff >= 0)
+    steps = diff . div step
+    exact_fit = case increasing of
+        True  -> start + Period.new months=steps*step == end
+        False -> start - Period.new months=steps*step == end
+    if exact_fit then steps else steps+1
```

Comment:
I was a bit skeptical about this heuristic, but I added as many test cases trying to breaking as I could imagine and it works, so I think we should be fine. The bright side is that it gives us O(1) time complexity for the case where the Period has `days=0`. So the brute force O(N) is now only when _both_ days and months are non-zero.

Apparently, the JDK uses a pretty smart way to compute the month difference:
```java
    private long monthsUntil(LocalDate end) {
        long packed1 = getProlepticMonth() * 32L + getDayOfMonth();  // no overflow
        long packed2 = end.getProlepticMonth() * 32L + end.getDayOfMonth();  // no overflow
        return (packed2 - packed1) / 32;
    }
```

(cited from [LocalDate.java](https://github.com/AdoptOpenJDK/openjdk-jdk12u/blob/master/src/java.base/share/classes/java/time/LocalDate.java#L1665-L1669))